fs/ref_client.go

```diff
@@ -0,0 +1,111 @@
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package fs
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/awslabs/soci-snapshotter/soci"
+	"github.com/containerd/containerd/platforms"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactsspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/registry/remote"
+)
+
+var (
+	ErrNoReferrers = errors.New("no existing referrers")
+)
+
+/// Determines which index will be selected from a list of index descriptors
+type IndexSelectionPolicy func([]ocispec.Descriptor) (ocispec.Descriptor, error)
+
+func SelectFirstPolicy(descs []ocispec.Descriptor) (ocispec.Descriptor, error) {
+	return descs[0], nil
+}
+
+/// Responsible for making Referrers API calls to remote registry to fetch list of referrers.
+type ReferrersClient interface {
+	/// Takes in an manifest descriptor and IndexSelectionPolicy and returns a single artifact descriptor.
+	/// Returns an error (ErrNoReferrers) if the manifest descriptor has no referrers.
+	Get(context.Context, ocispec.Descriptor, IndexSelectionPolicy) (ocispec.Descriptor, error)
+}
+
+type orasReferrersClient struct {
+	remote *remote.Repository
+}
+
+func NewORASReferrersClient(remote *remote.Repository) ReferrersClient {
+	return &orasReferrersClient{
+		remote: remote,
+	}
+}
+
+func (f *orasReferrersClient) Get(ctx context.Context, desc ocispec.Descriptor, fn IndexSelectionPolicy) (ocispec.Descriptor, error) {
+	descs, err := f.fetch(ctx, desc)
+	if err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("unable to fetch referrers: %w", err)
+	}
+	if len(descs) == 0 {
+		return ocispec.Descriptor{}, ErrNoReferrers
+	}
+	return fn(descs)
+}
+
+func (f *orasReferrersClient) fetch(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	descs := []ocispec.Descriptor{}
+	err := f.remote.Referrers(ctx, desc, soci.SociIndexArtifactType, func(referrers []artifactsspec.Descriptor) error {
+		for _, v := range referrers {
+			descs = append(descs, artifactToOCIDesc(v))
+		}
+		return nil
+	})
+	return descs, err
+}
+
+func artifactToOCIDesc(ar artifactsspec.Descriptor) ocispec.Descriptor {
+	return ocispec.Descriptor{
+		MediaType:   ar.MediaType,
+		Digest:      ar.Digest,
+		Size:        ar.Size,
+		Annotations: ar.Annotations,
+		URLs:        ar.URLs,
+	}
+}
+
+// Retrieves the manifest list from store and returns a manifest descriptor for the current platform
+func getManifestFromIndex(ctx context.Context, store content.Storage, desc ocispec.Descriptor, platform ocispec.Platform) (ocispec.Descriptor, error) {
```

Comment:
`soci.GetImageManifestDescriptor` is very similar to this. We should have a follow up to dedupe them.